tests/test_misc.lua

```diff
@@ -541,4 +542,382 @@ T['zoom()']['respects `config` argument'] = function()
+  eq(stat.type, 'file')
+  eq(stat.size > 1, true)
+end
+
+local read_shada_file = function()
+  -- verify that shada file exists
+  local stat = vim.loop.fs_stat(cursor_shada_file)
+  eq(stat.type, 'file')
+  eq(stat.size > 1, true)
+
+  child.cmd('rshada! ' .. cursor_shada_file)
+end
+
+T['restore_cursor()']['file'] = function()
+  local opts = restore_cursor_default_opts
+
+  child.cmd('edit ' .. test_file_cursor)
+
+  local jump_line = 23
+  cursor_jump_line(jump_line)
+  write_shada_file()
+  child.cmd('bdelete')
+
+  read_shada_file()
+  child.cmd('edit ' .. test_file_cursor)
+  eq(child.fn.line('.'), 1)
+  restore_cursor(opts)
+  eq(child.fn.line('.'), jump_line)
+end
+
+T['restore_cursor()']['help'] = function()
+  local opts = restore_cursor_default_opts
+
+  child.cmd('help')
+
+  local jump_line = 24
+  cursor_jump_line(jump_line)
+  write_shada_file()
+
+  child.cmd('quit')
+  read_shada_file()
+  child.cmd('help')
+  eq(child.fn.line('.'), 1)
+  restore_cursor(opts)
+  eq(child.fn.line('.'), jump_line)
+end
+
+T['restore_cursor()']['ignore_buftype'] = function()
+  local opts = {
+    ignore_buftype = { 'help' },
+    ignore_filetype = {},
+    center = true,
+  }
+
+  child.cmd('help')
+
+  local jump_line = 25
+  cursor_jump_line(jump_line)
+  write_shada_file()
+  child.cmd('quit')
+
+  read_shada_file()
+  child.cmd('help')
+  restore_cursor(opts)
+  eq(child.fn.line('.'), 1)
+end
+
+T['restore_cursor()']['ignore_filetype'] = function()
+  local opts = {
+    ignore_filetype = { 'lua' },
+    ignore_buftype = {},
+    center = true,
+  }
+
+  child.cmd('edit ' .. test_file_cursor)
+
+  local jump_line = 26
+  cursor_jump_line(jump_line)
+  write_shada_file()
+  child.cmd('bdelete')
+
+  read_shada_file()
+  child.cmd('edit ' .. test_file_cursor)
+  restore_cursor(opts)
+  eq(child.fn.line('.'), 1)
+end
+
+T['restore_cursor()']['folds'] = function()
+  local opts = restore_cursor_default_opts
+
+  child.cmd('edit ' .. test_file_cursor)
+
+  local jump_line = 15
+  cursor_jump_line(jump_line)
+  child.cmd('2,30 fold') -- fold test1 function
+  -- inside fold
+  eq(child.fn.foldclosed('10'), 2)
+  -- outside fold
+  eq(child.fn.foldclosed('40'), -1)
+  write_shada_file()
+  child.cmd('bdelete')
+
+  read_shada_file()
+  child.cmd('edit ' .. test_file_cursor)
+  eq(child.fn.line('.'), 1)
+  restore_cursor(opts)
+  eq(child.fn.line('.'), jump_line)
+  eq(child.fn.foldclosed('.'), -1)
+  eq(child.fn.foldclosed(10), -1)
+end
+
+T['restore_cursor()']['center'] = function()
+  local opts = restore_cursor_default_opts
+
+  child.cmd('edit ' .. test_file_cursor)
+
+  local jump_line = 32
+  cursor_jump_line(jump_line)
+  write_shada_file()
+  child.cmd('bdelete')
+
+  read_shada_file()
+  child.cmd('edit ' .. test_file_cursor)
+  restore_cursor(opts)
+  eq(child.fn.line('.'), jump_line)
+
+  local win_height = child.api.nvim_win_get_height(0)
+  local win_cursor = child.fn.winline()
+  eq(math.floor(win_height / 2 + 1), win_cursor)
```

Comment:
More explicit and robust approach is to set set of child process in the test start (`child.set_size(10, 12)` for example) and then test for top window line (`eq(child.fn.line('w0'), x)`)

Reply:
Done.